master/internal/api_trials.go

```diff
@@ -368,6 +370,292 @@ func (a *apiServer) GetTrialCheckpoints(
 	return resp, a.paginate(&resp.Pagination, &resp.Checkpoints, req.Offset, req.Limit)
 }
 
+func checkTrialFiltersEmpty(f *apiv1.TrialFilters) error {
+	emptyFilters := status.Errorf(
+		codes.InvalidArgument,
+		"at least one filter required",
+	)
+
+	if f == nil {
+		return emptyFilters
+	}
+
+	filtersLength := len(f.ExperimentIds) +
+		len(f.ProjectIds) +
+		len(f.WorkspaceIds) +
+		len(f.ValidationMetrics) +
+		len(f.TrainingMetrics) +
+		len(f.Hparams) +
+		len(f.Searcher) +
+		len(f.UserIds) +
+		len(f.Tags) +
+		len(f.States) +
+		len(f.SearcherMetric)
+
+	if filtersLength == 0 && 
+		f.RankWithinExp == nil &&
+	 	f.StartTime == nil && f.EndTime == nil && 
+	 	f.SearcherMetricValue == nil {
+			return emptyFilters
+	}
+	return nil
+}
+
+func (a *apiServer) QueryTrials(ctx context.Context, 
+	req *apiv1.QueryTrialsRequest) (*apiv1.QueryTrialsResponse, error) {
+	err := checkTrialFiltersEmpty(req.Filters)
+	if err != nil {
+		return nil, fmt.Errorf("error querying tags for trials %w", err)
+	}
+
+	trials := []db.TrialsAugmented{}
+
+	q := db.Bun().NewSelect().Model(&trials)
+	q, err = a.m.db.FilterTrials(q, req.Filters, true)
```

Comment:
how about moving `q := db.Bun().NewSelect().Model(&trials)` *inside* `BuildFilterTrialsQuery` 

i.e. in `BuildFilterTrialsQuery` do
```
q := db.Bun().NewSelect().Model((*db.TrialsAugmented)(nil))
```
and here do
```
q := BuildFilterTrialsQuery(req.Filters, true)
...
...
trials := []db.TrialsAugmented{}
q.Scan(context.TODO(), &trials)
```

this way the caller won't have to do this "Bun-NewSelect-Model" preamble for each `FilterTrials` call.

Reply:
yeah, this is a vestige from when `FilterTrials` took a `bun.QueryBuilder` (to accommodate `bun.UpdateQuery`). that ended up not working because `ExpRank` was doing a window expression in a `WHERE` clause, which apparently is not legal. so ended up doing `WHERE IN ({trial_id_subquery})` instead.

on the other hand the [update query](https://github.com/determined-ai/determined/pull/4543/files#diff-22336866e3c77178639f72a957328a898ce19c360034317e30275b87fb8869b6R482) is only using the`ids` from the filtered subquery, whereas  `q := db.Bun().NewSelect().Model(&trials)` would select them all. pretty sure the query compiler wouldnt bother, but i guess we would have to do an extra `q.select(id from ?', subQ)` [here](https://github.com/determined-ai/determined/pull/4543/files#diff-22336866e3c77178639f72a957328a898ce19c360034317e30275b87fb8869b6R482-R487). which to me is a little goofy, but maybe there is a better way..?

Reply:
won't this work?

```
q := BuildFilterTrialsQuery(req.Filters, true)
q = q.Column("trial_id")
```

Reply:
maybe so? will check. seems like bun generally gets confused when you specify things in two different ways (in this case, specifying the columns via both model and `.Column()`.

Reply:
yeah looks like that works.

 couple questions.

- Is there a name for this type of construct: `*db.TrialsAugmented)(nil)` and/or what does it do?
- in any case it looks like it avoids scanning into a model? but in the case of the select query we actually do want to scan. in which case we should just do a `.Scan(&here)` at the end?
- can you help me understand a little more the thinking behind where the boundaries should be between these functions? the initial thought was to have separate functions for different clauses: one handles `WHERE`, one handles `SET` etc,... although that also broke down around the `ExpRank` functionality....

Reply:
1. it's a typed nil pointer. it's typed so bun can reflect the type and figure out which model/table to query
2. yes
3. I'd break it down based on how you reuse and compose them around the code. we could come up with some naming scheme, like `Build...` creates a new query while `Apply...` modifies a query passed via parameter.